src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -108,10 +125,20 @@ where
                 mods_locked,
                 group,
             } => {
-                if !data.has_keymap.load(Ordering::Relaxed) {
-                    virtual_keyboard.post_error(NoKeymap, "`modifiers` sent before keymap.")
-                }
+                // Ensure keymap was initialized.
+                let virtual_data = data.handle.inner.lock().unwrap();
+                let keymap = match &virtual_data.keymap {
+                    Some(keymap) => keymap,
+                    None => {
+                        virtual_keyboard.post_error(NoKeymap, "`modifiers` sent before keymap.");
+                        return;
+                    }
+                };
+
+                // Ensure virtual keyboard's keymap is active.
                 let keyboard_handle = data.seat.get_keyboard().unwrap();
+                keyboard_handle.change_keymap(keymap);
```

Comment:
Just to be 100% clear, I wasn't actually able to verify `Modifiers` and `Key` requests in virtual keyboard still work, because the virtual keyboard I use don't use this (instead it inputs text through IME I believe).

If this would trigger `KeyboardHandle::input` this wouldn't work since that function would reset the keymap to the normal keyboard's keymap again. But it seems like this goes straight to the client so it should be fine?